src/math/binary-expression.ts

```diff
@@ -0,0 +1,160 @@
+    } else if (ast?.type === "number") {
+      return stringToBinaryNumber(ast.value) ?? ["Error", `Expected a binary number: ${ast.value}`];
+    } else {
+      return ["Error", `Expected binary number: ${JSON.stringify(ast)}`];
+    }
+  }
+
+  function customPrinter(value: MathJson<BinaryNumber>): string | null {
+    if (value instanceof BinaryNumber) {
+      return binaryNumberToLatex(value, opts);
+    } else {
+      return null;
+    }
+  }
+
+  function stringToBinaryNumber(value: string): BinaryNumber | null {
+    const binaryNumberRegex = /([+-])?([0-1]+)([.,]([0-1]+))?/;
+
+    const matchResults = (value ?? "").match(binaryNumberRegex);
+    if (matchResults === null) return null;
+
+    const [_, sign, numberValue, __, fractionalPart] = matchResults;
+
+    return new BinaryNumber(
+      sign === "-",
+      ((numberValue ?? "") + (fractionalPart ?? "")).split("").map((v) => (v === "0" ? false : true)),
+      fractionalPart?.length ?? 0
+    );
+  }
+
+  function binaryNumberToLatex(value: BinaryNumber, options?: BinaryNumberPrintingOptions) {
+    let printOptions = {
+      showSign: "minus",
+      zeroWidthDecimal: false,
+      groupZeros: true,
+      ...options,
+    };
+
+    // TODO: Currently unused, we could deprecate/remove the showSign option
+    const sign = printOptions.showSign === "never" ? "" : value.isNegative ? "-" : printOptions.showSign === "always" ? "+" : "";
+
+    let beforeDecimal = value
+      .getValueBeforeDecimal()
+      .map((v) => (v ? "1" : "0"))
+      .join("");
+
+    let afterDecimal = value
+      .getValueAfterDecimal()
+      .map((v) => (v ? "1" : "0"))
+      .join("");
+
+    if (printOptions.groupZeros) {
+      beforeDecimal = splitIntoChunks(beforeDecimal, 4, true).join("\\,");
+      afterDecimal = splitIntoChunks(afterDecimal, 4, false).join("\\,");
+    }
+
+    let output = `\\texttt{\\mathllap{${sign}}}` + `\\mathtt{${beforeDecimal}}`;
+
+    if (afterDecimal !== undefined && afterDecimal.length > 0) {
+      const decimalPoint = printOptions.zeroWidthDecimal ? "\\mathclap{\\raisebox{-0.1em}{.}}" : "\\mathtt{.}";
+
+      output += decimalPoint + `\\mathtt{${afterDecimal}}`;
+    }
+
+    return output;
+  }
+
+  function splitIntoChunks(value: string, chunkSize: number, backwards: boolean = false) {
+    const chunkCount = Math.ceil(value.length / chunkSize);
+    const chunks = new Array(chunkCount);
+
+    // If we have 100 and want the first block to be 4 bits long, we have to shift by -1
+    const nextLargestMultiple = chunkCount * chunkSize;
+    const startOffset = backwards ? value.length - nextLargestMultiple : 0;
+
+    for (let i = 0; i < chunks.length; i++) {
+      const position = i * chunkSize + startOffset;
+      chunks[i] = value.substring(position, position + chunkSize);
+    }
+
+    return chunks;
+  }
+
+  const mathJsonOperatorMap = new Map<MathJsonMathOperator, (a: BinaryNumber, b: BinaryNumber) => BinaryNumber>([
```

Comment:
can you explain to me the purpose of this map, because I don't really understand what it is supposed to do.

Reply:
I guess the values are irrelevant, but why would you use a complex type for the value then

Reply:
The idea was that one could give it some MathJson and this map would let you calculate the result. Though, I suppose I'll scrap those complicated types for now